fastn-core/src/auth/email_password.rs

```diff
@@ -0,0 +1,166 @@
+            fastn_core::schema::fastn_user_email::email
+                .eq(fastn_core::utils::citext(user_payload.email.as_str())),
+            fastn_core::schema::fastn_user_email::verified.eq(false),
+            fastn_core::schema::fastn_user_email::primary.eq(true),
+        ))
+        .execute(&mut conn)
+        .await?;
+
+    tracing::info!("fastn_user created. user_id: {}", &user.id);
+
+    if affected > 0 {
+        tracing::info!("fastn_user_email created");
+    }
+
+    fastn_core::http::api_ok(user)
+}
+
+pub(crate) async fn login(
+    req: &fastn_core::http::Request,
+    db_pool: &fastn_core::db::PgPool,
+    next: String,
+) -> fastn_core::Result<fastn_core::http::Response> {
+    use diesel::prelude::*;
+    use diesel_async::RunQueryDsl;
+
+    if req.method() != "POST" {
+        return Ok(fastn_core::not_found!("invalid route"));
+    }
+
+    #[derive(serde::Deserialize, Debug)]
+    struct Payload {
+        // TODO: add support for login using email
+        username: String,
+        password: String,
+    }
+
+    let payload: Payload = req.json()?;
+
+    let mut conn = db_pool
+        .get()
+        .await
+        .map_err(|e| fastn_core::Error::DatabaseError {
+            message: format!("Failed to get connection to db. {:?}", e),
+        })?;
+
+    let user: Option<fastn_core::auth::FastnUser> = fastn_core::schema::fastn_user::table
+        .filter(fastn_core::schema::fastn_user::username.eq(&payload.username))
+        .select(fastn_core::auth::FastnUser::as_select())
+        .first(&mut conn)
+        .await
+        .optional()?;
+
+    if user.is_none() {
+        return fastn_core::http::api_error(
+            "invalid payload",
+            fastn_core::http::StatusCode::BAD_REQUEST.into(),
+        );
+    }
+
+    let user = user.expect("already checked for None");
+
+    // OAuth users don't have password
+    if user.password.is_empty() {
+        // TODO: create feature to ask if the user wants to convert their account to an email
+        // password
+        // or should we redirect them to the oauth provider they used last time?
+        // redirecting them will require saving the method they used to login which de don't atm
+        return fastn_core::http::api_error(
+            "use available oauth providers to sign in",
+            fastn_core::http::StatusCode::BAD_REQUEST.into(),
+        );
+    }
+
+    let parsed_hash = argon2::PasswordHash::new(&user.password)
+        .map_err(|e| fastn_core::Error::generic(format!("failed to parse hashed password: {e}")))?;
+
+    let password_match = argon2::PasswordVerifier::verify_password(
+        &argon2::Argon2::default(),
+        payload.password.as_bytes(),
+        &parsed_hash,
+    );
+
+    if password_match.is_err() {
+        return fastn_core::http::api_error(
+            "incorrect username/password",
+            fastn_core::http::StatusCode::BAD_REQUEST.into(),
+        );
```

Comment:
The error handling is still not correct.

Reply:
We have to return something that fastn can consume: https://github.com/fastn-stack/fastn/blob/main/fastn-js/js/ftd.js#L218C45-L229

Reply:
There are bunch of compilation errors that are still need to be resolved (related to gh auth and the `user-details` processor). This PR is still a WIP. I'll definietly write proper json responses for error states once those errors are resolved.

Tho I think, ["incorrect username/password" is ok](https://superuser.com/a/1048752) as we do not want to let the client know what exactly went wrong, do we?